io.c

```diff
@@ -2244,6 +2237,184 @@ static int io_sqe_files_register(struct io_ring_ctx *ctx, void __user *arg,
 	return ret;
 }
 
+static int io_sqe_register_buffers(struct io_ring_ctx *ctx, void __user *uarg)
+{
+	struct vm_area_struct **vmas = NULL;
+	struct page **pages = NULL;
+	int buf_index, j;
+	int ret = -EINVAL;
+
+	struct io_mapped_ubuf *imu;
+	unsigned long off, start, end, ubuf;
+	int pret, nr_pages, max_nr_pages;
+	struct pxd_ioc_register_buffers arg;
+	size_t size;
+	bool existing = true;
+	long offset = 0;
+
+	ret = -EFAULT;
+	if (copy_from_user(&arg, uarg, sizeof(arg)))
+		goto err;
+
+	ret = -EINVAL;
+	if (arg.len & ~PAGE_MASK) {
+		pr_err("%s: mapped len is not aligned", __func__);
+		goto err;
+	}
+
+	if ((uintptr_t)arg.base & ~PAGE_MASK) {
+		pr_err("%s: mapped addr is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len & ~PAGE_MASK) {
+		pr_err("%s: mapped max len is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len < arg.len) {
+		pr_err("%s: max len less than len", __func__);
+		goto err;
+	}
+
+	for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+		imu = &ctx->user_bufs[buf_index];
+		if (imu->ubuf <= (uintptr_t) arg.base &&
+		    (uintptr_t) arg.base + arg.len <= imu->ubuf + imu->len) {
+			offset = ((uintptr_t)arg.base - imu->ubuf) >> PAGE_SHIFT;
+			pr_info("found existing mapping %d, off %ld", buf_index, offset);
+			break;
+		}
+		if (imu->ubuf < (uintptr_t) arg.base + arg.len &&
+			(uintptr_t) arg.base < imu->ubuf + imu->len) {
+			pr_info("%s: intersects with existing mapping", __func__);
+			ret = -EEXIST;
+			goto err;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		existing = false;
+		for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+			pr_info("new mapping %d, ", buf_index);
+			if (ctx->user_bufs[buf_index].ubuf == 0)
+				break;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		pr_err("%s: all busy", __func__);
+		return -EBUSY;
+	}
+
+	imu = &ctx->user_bufs[buf_index];
+
+	/*
+         * Don't impose further limits on the size and buffer
+         * constraints here, we'll -EINVAL later when IO is
+         * submitted if they are wrong.
+         */
+	ret = -EFAULT;
+	if (!arg.base || !arg.len)
+		goto err;
+
+	/* arbitrary limit, but we need something */
+	if (arg.len > SZ_4G)
+		goto err;
+
+	ubuf = (unsigned long) arg.base;
+	end = (ubuf + arg.len) >> PAGE_SHIFT;
+	start = ubuf >> PAGE_SHIFT;
+	nr_pages = end - start;
+	max_nr_pages = arg.max_len >> PAGE_SHIFT;
+
+	if (existing) {
+		for (j = 0; j < nr_pages; ++j) {
+			if (imu->bvec[j + offset].bv_page) {
```

Comment:
imu->bvec assumed to be available if existing.
But if there is any map failure, the entire bvec is freed. So it opens up possibility for bvec to be NULL, but the mapping range to exist.